zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -723,6 +913,459 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
             ]
         });
 
+        // Block hash checks in three parts:
+        // 1. RLP checks
+        // 2. RLC calculation
+        // 3. Keccak lookup
+
+        // 1. Block header RLP
+        meta.lookup_any("Block header RLP: byte range checks", |meta| {
+            let block_header_rlp = meta.query_advice(blk_hdr_rlp, Rotation::cur());
+            let fixed_u8_table = meta.query_fixed(fixed_u8, Rotation::cur());
+
+            vec![(block_header_rlp, fixed_u8_table)]
+        });
+
+        meta.create_gate("Block header RLP: constant checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+
+            let q_blk_hdr_rlp = meta.query_selector(q_blk_hdr_rlp);
+            let q_blk_hdr_rlp_const = meta.query_selector(q_blk_hdr_rlp_const);
+            let blk_hdr_rlp = meta.query_advice(blk_hdr_rlp, Rotation::cur());
+            let blk_hdr_rlp_const = meta.query_fixed(blk_hdr_rlp_const, Rotation::cur());
+
+            cb.require_equal(
+                "RLP hdr costants are correct",
+                blk_hdr_rlp,
+                blk_hdr_rlp_const,
+            );
+            cb.gate(and::expr([q_blk_hdr_rlp, q_blk_hdr_rlp_const]))
+        });
+
+        // Make sure that length starts from 0
+        meta.create_gate("Block header RLP: length default value = 0", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+
+            let length = meta.query_advice(blk_hdr_rlp_len_calc, Rotation::cur());
+            cb.require_zero("length default value is zero", length);
+
+            cb.gate(and::expr([
+                not::expr(or::expr([
+                    meta.query_fixed(q_number, Rotation::cur()),
+                    meta.query_fixed(q_var_field_256, Rotation::cur()),
+                ])),
+                meta.query_selector(q_blk_hdr_rlp),
+            ]))
+        });
+
+        meta.create_gate(
+            "Block header RLP: leading zeros column is boolean",
```

Comment:
So the reason I am adding leading zeros is only for padding so that the RLP data fit the right rows. The leading zeros are not taken into account in the RLP itself.